### PR TITLE
Add Global Retention Policy Agent for DB

### PR DESCRIPTION
**Pull request by khrm**
Add Result Retention Policy Agent which removes older Result and
their associated records. Retention Policy is based on the
configmap config-results-retention-policy.yaml

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you review them:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] [Tested your changes locally](https://github.com/tektoncd/results/blob/main/docs/DEVELOPMENT.md) (if this is a code change)
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [x] Has a kind label. You can add a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user-facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contain the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Add Result Retention Policy Agent which removes older Result and their associated records.
```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/kind feature

**Review by khrm — COMMENTED** [review]
/assign @vdemeester @enarha Can you please look into this?

File: cmd/retention-policy-agent/main.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	"github.com/tektoncd/results/pkg/retention"
+	_ "go.uber.org/automaxprocs"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
+	_ "knative.dev/pkg/client/injection/kube/client"
+
+	gormlogger "gorm.io/gorm/logger"
+	"k8s.io/apimachinery/pkg/util/wait"
```
Comment (vdemeester) [inline cmd/retention-policy-agent/main.go:37]: nit: these import could be re-organized 🙃

**Review by vdemeester — APPROVED** [review]

File: cmd/retention-policy-agent/main.go (hunk 1 above)
Comment (khrm) [reply cmd/retention-policy-agent/main.go]: Re-Organised them.

**Comment by khrm** [comment]
Checking for flakiness by retesting again. All two previous runs passed. 
/test pull-tekton-results-unit-tests

File: docs/retention-policy-agent/README.md [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+---
+linkTitle: "Results Retention Policy Agent"
+weight: 2
+---
+
+-->
+
+# Result Retention Policy Agent
+
+The Result Retention Policy Agent removes older Result and their associated records.
```
[the remark this answers is not in the document]
Comment (khrm) [reply docs/retention-policy-agent/README.md]: Done.

File: docs/retention-policy-agent/README.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+## Configuration
+The Agent config uses following field
```
[the remark this answers is not in the document]
Comment (khrm) [reply docs/retention-policy-agent/README.md]: Done.

File: pkg/apis/config/retention.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	return newRetentionPolicyFromMap(config.Data)
+}
+
+// GetRetentionPolicyConfigName returns the name of the configmap containing
+// retention policy.
+func GetRetentionPolicyConfigName() string {
+	if e := os.Getenv("CONFIG_RETENTION_POLICY_NAME"); e != "" {
+		return e
+	}
+	return "tekton-results-config-results-retention-policy"
```
[the remark this answers is not in the document]
Comment (khrm) [reply pkg/apis/config/retention.go]: Done.

File: config/base/config-results-retention-policy.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: config-results-retention-policy
+  labels:
+    app.kubernetes.io/name: tekton-results-logging
```
[the remark this answers is not in the document]
Comment (khrm) [reply config/base/config-results-retention-policy.yaml]: Done.

**Comment by khrm** [comment]
I have addressed all review comments @enarha

**Comment by angeliski** [comment]
Hey @khrm, thanks for your amazing work in this PR!

Two questions about that:
1. We can avoid delete old records just not starting the agent, right? Maybe we can put that in the docs (I can raise a PR if you think that make sense)

2. I just update my watcher to v0.13.2 and start to receive the error: 
```
,"msg":"Failed to start configuration manager","commit":"683497f-dirty","er │
│ ror":"configmap \"tekton-results-config-results-retention-policy\" not found","stacktrace":"knative.dev/pkg/injection/sharedmain.MainWithConfig\n\tknative.dev/pkg@v0.0.0-20240614135239-339c22b8218c/injection/sharedmain │
│ /main.go:285\nmain.main\n\tgithub.com/tektoncd/results/cmd/watcher/main.go:151\nruntime.main\n\truntime/proc.go:271"}

```

I read the PR and imagine this is just a side effect not desired, right? Should I open an issue for that?

**Pull request merged**